### PR TITLE
Keep connection opened by ping

### PR DESCRIPTION
**Pull request by matejvasek**
`podman` (alternative daemon docker dameon)  is closing connections after 20 seconds without any request. This is causing troubles when `build phase` takes longer than that and builder is trusted.

If builder is trusted then docker client's connection is established at the benign while process is still uid zero. Then the connection is supposed to be kept opened. However `podman` is closing it.

This commit sends `ping` request to the daemon each second while `build phase` is in progress.

**Steps to reproduce the issue**:
1. create simple Java app: `mvn archetype:generate -DarchetypeGroupId=org.apache.maven.archetypes -DarchetypeArtifactId=maven-archetype-quickstart -DarchetypeVersion=1.4 -DgroupId=com.example -DartifactId=test-bp -B`
2. in one terminal run `podman` as a service: `podman  system service unix:///tmp/podman.sock --time=0 --log-level=debug`
3. in another terminal set docker socket via envvar: `export DOCKER_HOST="unix:///tmp/podman.sock"` and then try to build the app: `pack build test-bp --builder paketobuildpacks/builder:full --docker-host=inherit --trust-builder=true --pull-policy=if-not-present -v`

The build fails:
```
ERROR: failed to  image: error during connect: Get "http://%2Fvar%2Frun%2Fdocker.sock/v1.40/info": EOF
ERROR: failed to build: executing lifecycle: failed with status code: 1
```
You can also noticed following in the `podman`'s log:
```
DEBU[0045] IdleTracker 0xc000590010:closed 3m+1h/4t connection(s) 
DEBU[0045] IdleTracker 0xc0000111f0:closed 2m+1h/4t connection(s) 
```

### CODE REVIEW

**Comment by aemengo** [comment]
Is this `ping` request something that the docker daemon presently does? I'm just _slightly_ perturbed about how idiosyncratic this request feels.

Can I also ask the problem that podman was trying to solve with this connection timeout?

**Comment by matejvasek** [comment]
@aemengo I know this looks awful. I also filled PR https://github.com/containers/podman/pull/11072 
This is kinda workaround patching another workaround IMHO. The container is deescalating rights at some point, that leads to possible inability to dial unix socket. So the socket is dialed at beginning (while uid=0) and it should be kept open. The thing is that `docker` is not automatically closing unused connections while `podman` does close unused connection after 20sec.

**Comment by matejvasek** [comment]
>The thing is that docker is not automatically closing unused connections

Also I am not sure if one can relay on this.

**Comment by matejvasek** [comment]
I don't like that the connection is established once on start while running as a root. It shouldn't be needed to be root to dial socket. There is already the `docker` group for that.

So maybe better solution would be to add new flag for `lifecycle` binaries called `--docker-gid`.
Then `pack` could set the flag to numeric id of the `docker` group on given machine. `lifecycle` would call `func Setgroups(gids []int) (err error) ` with the value. This would allow non-root user to dial docker unix socket.

This is however much larger change than the `ping hack`, this would change public API.

wdyt @aemengo

**Comment by matejvasek** [comment]
maybe it wouldn't need new flag, it could get the gid from the mounted socket  via stat

**Comment by aemengo** [comment]
> I don't like that the connection is established once on start while running as a root.

I've never been a stickler for permissions in ephemeral containers, but that's my failing.

> Then the connection is supposed to be kept opened.

Pardon my ignorance, but why must this be the case? From what I see here, ~Thanos~ podman did nothing wrong. Lifecycle should close its connection after use and then _re-dial_ when it needs to talk to the daemon. That's _imho_.

--

Regardless, I think this project has a bad habit of fixing things without reproduction cases. Perhaps a small reproduction case (_for example: a build phase failing on podman_) in your PR would help us get on the same page.

**Comment by matejvasek** [comment]
>Lifecycle should close its connection after use and then re-dial (re-escalate, if needed) when it needs to talk to the daemon.

Totally agree but I am not sure if it's possible to easily and safely elevate right.

Beside using `docker group` I can think of another solution: spawn process under root that would basically work as `socat`, the process would create new docker socket (with non-root access alowed) then the process would just forwarded request to the real docker socket.

**Comment by matejvasek** [comment]
@aemengo I added steps to reproduce.

**Comment by matejvasek** [comment]
cc @ekcasey

**Comment by matejvasek** [comment]
@natalieparellano @aemengo is it conceivable to reopen and merge this PR? The issue with the connection being closed surfaced again.

**Comment by matejvasek** [comment]
@natalieparellano @aemengo PTAL

**Review by jabrown85 — APPROVED** [review]

File: cmd/lifecycle/creator.go [hunk 1]
```diff
@@ -247,6 +268,8 @@ func (c *createCmd) Exec() error {
 		platform:      c.platform,
 		platformDir:   c.platformDir,
 	}.build(group, plan)
+	cancelPing()
+	<-pingDoneChan
```
Comment (natalieparellano) [inline cmd/lifecycle/creator.go:272]: Is there a need to block here? If we don't need the channel, that would make the code a bit simpler.

File: cmd/lifecycle/creator.go [hunk 2]
```diff
@@ -239,6 +241,25 @@ func (c *createCmd) Exec() error {
...
+		}
+		for {
+			select {
+			case <-time.After(time.Second * 1):
+				_, _ = c.docker.Ping(pingCtx)
+			case <-pingCtx.Done():
+				return
+			}
+		}
+	}()
```
Comment (natalieparellano) [inline cmd/lifecycle/creator.go:261]: Could we wrap this in a named function, to make the code easier to read? I think something like the below would do the trick (I left out the channel as I didn't see the need for it, but maybe I'm missing something):
```
func (c *createCmd) startPing(pingCtx context.Context) {
	if c.docker == nil {
		return
	}

	go func() {
		for {
			select {
			case <-time.After(time.Second * 1):
				_, _ = c.docker.Ping(pingCtx)
			case <-pingCtx.Done():
				return
			}
		}
	}()
}
```

**Comment by natalieparellano** [comment]
Thank you @matejvasek!

I was able to test this out with podman and it looks good. I left a couple code suggestions.

File: cmd/lifecycle/creator.go (hunk 1 above)
Comment (matejvasek) [reply cmd/lifecycle/creator.go:272]: It's not necessary but it shouldn't block for long, I just like to wait for goroutine if possible.

Comment (matejvasek) [reply cmd/lifecycle/creator.go:272]: It should be pretty instantaneous.

**Comment by matejvasek** [comment]
how about:
```diff
diff --git a/cmd/lifecycle/creator.go b/cmd/lifecycle/creator.go
index 0b960754..f648fd12 100644
--- a/cmd/lifecycle/creator.go
+++ b/cmd/lifecycle/creator.go
@@ -242,24 +242,7 @@ func (c *createCmd) Exec() error {
 		}
 	}
 
-	pingCtx, cancelPing := context.WithCancel(context.Background())
-	defer cancelPing()
-	pingDoneChan := make(chan struct{}, 1)
-	// send pings to docker daemon while BUILDING to prevent connection closure
-	go func() {
-		defer func() { pingDoneChan <- struct{}{} }()
-		if c.docker == nil {
-			return
-		}
-		for {
-			select {
-			case <-time.After(time.Millisecond * 500):
-				_, _ = c.docker.Ping(pingCtx)
-			case <-pingCtx.Done():
-				return
-			}
-		}
-	}()
+	stopPing := c.startPing()
 
 	cmd.DefaultLogger.Phase("BUILDING")
 	err = buildArgs{
@@ -269,8 +252,9 @@ func (c *createCmd) Exec() error {
 		platform:      c.platform,
 		platformDir:   c.platformDir,
 	}.build(group, plan)
-	cancelPing()
-	<-pingDoneChan
+
+	stopPing()
+
 	if err != nil {
 		return err
 	}
@@ -337,3 +321,29 @@ func (c *createCmd) populateRunImage() error {
 	}
 	return nil
 }
+
+func (c *createCmd) startPing() (stopPing func()) {
+	pingCtx, cancelPing := context.WithCancel(context.Background())
+	defer cancelPing()
+	pingDoneChan := make(chan struct{}, 1)
+	// send pings to docker daemon while BUILDING to prevent connection closure
+	go func() {
+		defer func() { pingDoneChan <- struct{}{} }()
+		if c.docker == nil {
+			return
+		}
+		for {
+			select {
+			case <-time.After(time.Millisecond * 500):
+				_, _ = c.docker.Ping(pingCtx)
+			case <-pingCtx.Done():
+				return
+			}
+		}
+	}()
+
+	return func() {
+		cancelPing()
+		<-pingDoneChan
+	}
+}
```

**Comment by matejvasek** [comment]
@natalieparellano I believe we must wait for the goroutine.

**Comment by matejvasek** [comment]
@natalieparellano PTAL

**Review by natalieparellano — APPROVED** [review]

**Pull request merged**